# Notebook 4 ---- Walk-Forward Forecasting with LightGBM for Production

In this notebook we go a step further with LightGBM (the best scoring model from our benchmark) by implementing a walk-forward forecasting algorithm to get a better idea of its true performance on a multi-store setting.

## Imports

In [5]:
import numpy as np
import pandas as pd
print(pd.__version__)

import lightgbm as lgb

import warnings
warnings.filterwarnings('ignore')

2.3.3


## Data Loading

In [ ]:
train = pd.read_csv('../data/raw/train.csv', parse_dates=['Date'])
store = pd.read_csv('../data/raw/store.csv')

df = train.merge(store, on='Store', how='left')
df = df[(df['Open'] == 1) & (df['Sales'] > 0)].copy()
df = df.sort_values(['Store', 'Date']).reset_index(drop=True)
df["StateHoliday"] = df["StateHoliday"].astype(str)
# saving as .parquet for later use by the production script
df.to_parquet("../data/rossmann.parquet")

print(f"Shape : {df.shape}")
print(f"Nb stores : {df['Store'].nunique()}")
print(f"Period : {df['Date'].min()} → {df['Date'].max()}")

Shape : (844338, 18)
Nb stores : 1115
Period : 2013-01-01 00:00:00 → 2015-07-31 00:00:00


## Feature Engineering

### Averages by Window

In the following block we built a function for computing rolling averages and rolling average rations for our dataset. The function has the following key steps:
1. it takes as argument a dataset and a tuple of windows (the windows on which the rolling calculations will be performed)
2. it sorts the data by store and then by date (obviously required for computing rolling averages)
3. it groups the data by store to avoid any overflow from one store to another of any rolling average calculation
4. finally, the for loop computes the different rolling averages and ratios for each window. It works at the group level as already explained above.
5. it returns the input dataset enriched with the newly computed averages and ratios

**Key point:** we will have to make sure, when building our train and test sets, that those averages and ratios do not leak any data.

In [3]:
def build_asof_features(df, windows=(7, 28, 91)):
    """
    Aggregate sales for a given windows ending at the current date
    """
    # sorting by stores and dates
    d = df.sort_values(["Store", "Date"]).copy()
    # days closed must not impact the averages
    d["SalesOpen"] = d["Sales"].where(d["Open"] == 1)
    g = d.groupby("Store")["SalesOpen"]
    # each row of g is a store object containing all sales data

    out = d[["Store", "Date"]].copy()
    for w in windows:
        mp = max(3, w // 3)
        out[f"asof_mean_{w}d"] = g.transform(lambda s, w=w, mp=mp: s.rolling(w, min_periods=mp).mean())
        out[f"asof_std_{w}d"]  = g.transform(lambda s, w=w, mp=mp: s.rolling(w, min_periods=mp).std())
    
    out["asof_ratio_7_28"] = out["asof_mean_7d"] / out["asof_mean_28d"]
    out["asof_ratio_28_91"] = out["asof_mean_28d"] / out["asof_mean_91d"]
    out["asof_cv_28"] = out["asof_std_28d"]  / out["asof_mean_28d"]

    return out

### Origin and Target Combinations

The overall logic of the walk-forward algorithm is to look at different "origin" date, build the model with all data available at and before the origin and test the model on data that comes after the origin. Think about a real-life monthly forecasting cycle:
- in March, you have the data from March itself and the previous months to build your forecasts for the future (April, May, etc.)
- comes the April forecasting cycle and you now have to one extra month of actuals
- and so on...
As we explore more recent origins, we have more data to build our model on.

The "make_training_frame" function does the following:
- it takes a list of origins as argument. This list of origins is itself based on the cutoff we consider for our train/test split
    - to building the training frame we feed the function with a list of the origins prior to the cutoff
    - to build the testing frame, we use the cutoff origin itself as argument.
- to build the training frames the function will loop through the list of origins whereas it will only look at one origin to build the testing frame. However the logic is the same. It just materializes the fact that comes April we can train our model on more data than we had in March (see previous paragraph).
- for each origin in the for loop the function will look at the data points that comes **after** this origin but **before** the horizon considered.
- for each entry in this subset of data (called tgt in the function), the will add the "asof" information corresponding to the last known state of the store.
- finally, since for training we may have looked into several origins, the function concatenate the different tgt frames (to which was added the information about the origin considered as cutoff between train and test) into one final data frame.

In [4]:
def snapshot_at(asof, ori, cols):
    """
    Last known state of each store at the origin date (included)
    This to avoid removing stores just because they were close at origin
    """
    # Select all the date at the origin or prior to the origin data
    s = asof[asof["Date"] <= ori]
    # For each store, only keep the most recent instance
    # which will be an open day since we already fitlered out the closed days
    s = s.drop_duplicates("Store", keep="last")
    s = s.rename(columns={"Date": "snap_date"})
    # Below is a measure of the freshness of the data
    # How close to the origin the last available data point is
    s["days_since_snap"] = (ori - s["snap_date"]).dt.days
    return s[["Store", "days_since_snap"] + cols]


def make_training_frame(df, origins, horizon=42, asof=None):
    if asof is None:
        asof = build_asof_features(df)
    asof_cols = [c for c in asof.columns if c.startswith("asof_")]

    # sales from the previous year anchored on target date
    # 364 days or 52 weeks, same day of week
    # 364 > horizon so always observed at origin
    ly = df[["Store", "Date", "Sales"]].rename(columns={"Sales": "sales_ly"})
    # Shifting the date by +364 days for later merging
    ly["Date"] = ly["Date"] + pd.Timedelta(days=364)

    frames = []
    for ori in pd.to_datetime(sorted(origins)):
        # tgt dataframe corresponds to dates after the origin but not after the forecasting horizon
        tgt = df[(df["Date"] > ori) & (df["Date"] <= ori + pd.Timedelta(days=horizon))].copy()
        # making use of the previous function to identify the last known state of each store at origin
        snap = snapshot_at(asof, ori, asof_cols)
        # merging back the data from the last know snapshot to the main df
        # in each tgt df, "Sales" is the target variable
        # the algorithm will learn using the last data know at the origin
        tgt = tgt.merge(snap, on="Store", how="inner")
        tgt = tgt.merge(ly, on=["Store", "Date"], how="left")
        tgt["origin"] = ori
        tgt["h"] = (tgt["Date"] - ori).dt.days
        tgt["h_week"] = np.ceil(tgt["h"] / 7).astype(int)
        frames.append(tgt)

    if not frames:
        raise ValueError(f"no workable origin into {list(origins)}")
    out = pd.concat(frames, ignore_index=True)
    return out[out["Open"] == 1].reset_index(drop=True)

### Origins and Folds

This builds on what we just did before and what I wrote previously. Here is the overall logic:
- we list all possible origins (monthly origins here)
- a iterates through all possible origins:
    - make_training_frame is called twice per iteration:
        - once with all origins prior to the selected one => this is used to build the training set for this fold
        - once with the selected origin itself => this is used to build the testing set for this fold
    - **importantly**, the training set (tr) is filtered to remove all dates in the future from the perspective of the training set itself (which means all dates at or after the origin). Not doing this would lead to test date leakage into the training set.


In [5]:
HORIZON = 42

candidate_origins = pd.date_range(
    start=df["Date"].min() + pd.Timedelta(days=400),
    end=df["Date"].max() - pd.Timedelta(days=HORIZON),
    freq="14D")

test_origins = pd.date_range(
    end=df["Date"].max() - pd.Timedelta(days=HORIZON), periods=6, freq="MS"
)

asof = build_asof_features(df).sort_values(["Store", "Date"]).reset_index(drop=True) # computed once for all folds

def make_fold(ori, margin=True):
    cutoff = ori - pd.Timedelta(days=HORIZON) if margin else ori
    tr_origins = candidate_origins[candidate_origins < cutoff]
    tr = make_training_frame(df, tr_origins, HORIZON, asof)
    tr = tr[tr["Date"] < ori]
    te = make_training_frame(df, [ori], HORIZON, asof)
    assert tr["Date"].max() < ori < te["Date"].min(), f"overflow on {ori.date()}"
    return tr, te

tr, te = make_fold(test_origins[0])
print(tr.shape, te.shape, tr["Date"].max(), te["Date"].min())

(746294, 32) (39923, 32) 2014-12-24 00:00:00 2015-01-02 00:00:00


In [6]:
for ori in test_origins:
    tr, te = make_fold(ori)
    print(ori.date(), len(tr), len(te), tr["Date"].max().date(), te["Date"].min().date())

2015-01-01 746294 39923 2014-12-24 2015-01-02
2015-02-01 811698 40170 2015-01-21 2015-02-02
2015-03-01 887804 38090 2015-02-18 2015-03-02
2015-04-01 968138 37040 2015-03-18 2015-04-02
2015-05-01 1084595 37472 2015-04-29 2015-05-02
2015-06-01 1158695 39626 2015-05-27 2015-06-02


## Feature Selection & Preparation

The categorical features are encoded once for the whole dataset to ensure that, in each fold, **tr** and **te** follows the same rule.

In [7]:
import lightgbm as lgb

CATEGORICAL = ["Store", "StoreType", "Assortment", "StateHoliday", "DayOfWeek"]

FEATURES = [
    # Last known state at origin
    "asof_mean_7d", "asof_mean_28d", "asof_mean_91d", "asof_std_28d",
    "asof_ratio_7_28", "asof_ratio_28_91", "asof_cv_28", "days_since_snap",
    # horizon
    "h", "h_week",
    # same day the previous year
    "sales_ly",
    # date and promotion related features (known in advance)
    "DayOfWeek", "Promo", "SchoolHoliday", "StateHoliday", "month", "day", "weekofyear",
    # store-related features
    "Store", "StoreType", "Assortment", "CompetitionDistance", "Promo2",
]

CATS = {c: pd.Index(sorted(df[c].astype(str).unique())) for c in CATEGORICAL}

def prep(d):
    d = d.copy()
    d["month"] = d["Date"].dt.month
    d["day"] = d["Date"].dt.day
    d["weekofyear"] = d["Date"].dt.isocalendar().week.astype(int)
    for c in CATEGORICAL:
        d[c] = pd.Categorical(d[c].astype(str), categories=CATS[c])
    return d

## Evaluation Metrics

In [8]:
def wmape(y, yhat):
    return np.abs(y - yhat).sum() / y.sum()

def bias(y, yhat):
    return (yhat - y).sum() / y.sum()

## Training

In [44]:
# LightGBM hyperparameters
PARAMS = dict(
    objective="regression", metric="l2", learning_rate=0.05,
    num_leaves=127, min_data_in_leaf=100,
    feature_fraction=0.8, bagging_fraction=0.8, bagging_freq=1,
    verbosity=-1, n_jobs=-1, seed=42,
)

def run_backtest(features):
    results = []
    imp = []
    # for each origin
    for O in test_origins:
        # we build the tr and te sets
        tr, te = make_fold(O)
        tr, te = prep(tr), prep(te)

        # we train lgbm on tr
        dtrain = lgb.Dataset(tr[features], np.log1p(tr["Sales"]), categorical_feature=CATEGORICAL)
        model = lgb.train(PARAMS, dtrain, num_boost_round=600)
        # and use the model to predict on te
        te = te.assign(pred=np.expm1(model.predict(te[features])).clip(0))

        # results are store by h_week (i.e, # weeks after origin)
        # this to show that performance degrades as we move away from origin (hypothesis)
        # because our "asof" features are less prevalent the farther we are from the origin
        for hw, g in te.groupby("h_week"):
            m = g["sales_ly"].notna()
            results.append(dict(
                origin=O.date(), h_week=int(hw), n=len(g),
                wmape=wmape(g["Sales"], g["pred"]),
                bias=bias(g["Sales"], g["pred"]),
                wmape_ly=wmape(g.loc[m, "Sales"], g.loc[m, "sales_ly"]) if m.any() else np.nan,
                n_ly=int(m.sum()),
            ))

        feat_imp = pd.Series(model.feature_importance(importance_type="gain"), index=features).sort_values(ascending=False)
        imp.append(feat_imp)

    imp = pd.concat(imp, axis=1).mean(axis=1)

    return pd.DataFrame(results), imp

In [14]:
res_full, res_full_feat_imp = run_backtest(FEATURES)
# after a first run wil all features, we noticed the outmost importance of the sales_ly feature
# we then devised another experiment to see what would be the performance without this top feature
res_no_ly, res_no_ly_feat_imp = run_backtest([f for f in FEATURES if f != "sales_ly"])

## Results

We built a first model using all the features in the FEATURE variable and noticed that although the WMAPE was good, one feature, "sales_ly", was much more important than all the others. We then decided to build a new model without this feature just to see whether it was cannibalizing all the others. Below is the result comparison between the two verions:

### Results by Origin and Horizon

In [27]:
# with all the features
res_full.pivot(index="h_week", columns="origin", values="wmape").round(3)

origin,2015-01-01,2015-02-01,2015-03-01,2015-04-01,2015-05-01,2015-06-01
h_week,,,,,,
1,0.178,0.104,0.075,0.195,0.168,0.100
2,0.158,0.078,0.078,0.156,0.142,0.099
3,0.164,0.090,0.085,0.207,0.066,0.099
4,0.121,0.097,0.112,0.074,0.109,0.072
5,0.146,0.078,0.190,0.098,0.093,0.121
6,0.110,0.089,0.121,0.106,0.105,0.098


In [28]:
# without sales_ly
res_no_ly.pivot(index="h_week", columns="origin", values="wmape").round(3)

origin,2015-01-01,2015-02-01,2015-03-01,2015-04-01,2015-05-01,2015-06-01
h_week,,,,,,
1,0.192,0.120,0.071,0.205,0.120,0.093
2,0.200,0.085,0.073,0.138,0.128,0.099
3,0.178,0.088,0.071,0.171,0.064,0.083
4,0.181,0.089,0.081,0.070,0.104,0.078
5,0.184,0.076,0.198,0.090,0.086,0.124
6,0.150,0.077,0.111,0.096,0.086,0.089


### Comparison with Naive Baseline

In [30]:
# with all the features
summary = res_full.groupby("h_week").apply(
    lambda g: pd.Series({
        "wmape": np.average(g["wmape"], weights=g["n"]),
        "bias": np.average(g["bias"], weights=g["n"]),
        "wmape_naif_ly": np.average(g["wmape_ly"], weights=g["n"]),
    })
)
summary["fva"] = 1 - summary["wmape"] / summary["wmape_naif_ly"]
print(summary.round(4))

         wmape    bias  wmape_naif_ly     fva
h_week                                       
1       0.1337  0.0473         0.1251 -0.0687
2       0.1178  0.0378         0.1438  0.1808
3       0.1187  0.0608         0.1639  0.2758
4       0.0972 -0.0212         0.1454  0.3312
5       0.1202  0.0111         0.1359  0.1153
6       0.1044 -0.0124         0.1113  0.0624


In [31]:
# without sales_ly
summary = res_no_ly.groupby("h_week").apply(
    lambda g: pd.Series({
        "wmape": np.average(g["wmape"], weights=g["n"]),
        "bias": np.average(g["bias"], weights=g["n"]),
        "wmape_naif_ly": np.average(g["wmape_ly"], weights=g["n"]),
    })
)
summary["fva"] = 1 - summary["wmape"] / summary["wmape_naif_ly"]
print(summary.round(4))

         wmape    bias  wmape_naif_ly     fva
h_week                                       
1       0.1294  0.0342         0.1251 -0.0349
2       0.1205  0.0473         0.1438  0.1622
3       0.1090  0.0603         0.1639  0.3350
4       0.1002  0.0059         0.1454  0.3104
5       0.1261  0.0111         0.1359  0.0717
6       0.1013  0.0026         0.1113  0.0906


### WMAPE Comparison between Model 1 (all features) and Model 2 (without sales_ly)

In [15]:
comp = pd.concat({
    "with_sales_ly": res_full.groupby("h_week").apply(lambda g: np.average(g["wmape"], weights=g["n"])),
    "without_sales_ly": res_no_ly.groupby("h_week").apply(lambda g: np.average(g["wmape"], weights=g["n"])),
}, axis=1)
comp["delta"] = comp["without_sales_ly"] - comp["with_sales_ly"]
print(comp.round(4))

        with_sales_ly  without_sales_ly   delta
h_week                                         
1              0.1337            0.1294 -0.0042
2              0.1178            0.1205  0.0027
3              0.1187            0.1090 -0.0097
4              0.0972            0.1002  0.0030
5              0.1202            0.1261  0.0059
6              0.1044            0.1013 -0.0031


### Feature Importance

#### Model with all features

In [18]:
res_full_feat_imp.head(10)

sales_ly         816960.596528
asof_mean_28d    174754.748984
Promo            166463.772805
asof_mean_91d    163800.013783
Store            115809.901374
DayOfWeek         57418.982777
day               41886.585610
weekofyear        40429.769735
month             19640.772141
asof_mean_7d       8570.132645
dtype: float64

#### Model without sales_ly

In [19]:
res_no_ly_feat_imp.head(10)

asof_mean_91d    427765.072715
asof_mean_28d    357778.163075
Promo            267806.744728
Store            170708.070798
asof_mean_7d     118973.849729
DayOfWeek        102780.888137
day               57223.222728
weekofyear        46728.365016
month             24086.568406
asof_cv_28        19753.510279
dtype: float64

#### Feature Ranking Comparison

In [26]:
pd.concat({"with_sales_ly": res_full_feat_imp.rank(ascending=False), "without_sales_ly": res_no_ly_feat_imp.rank(ascending=False)}, axis=1)

,with_sales_ly,without_sales_ly
sales_ly,1.0,NaN
asof_mean_28d,2.0,2.0
Promo,3.0,3.0
asof_mean_91d,4.0,1.0
Store,5.0,4.0
DayOfWeek,6.0,6.0
day,7.0,7.0
weekofyear,8.0,8.0
month,9.0,9.0
asof_mean_7d,10.0,5.0


## Conclusion

### Methodology

**Multi-Horizon Direct Forecasting**: one single model predicts [D+1, D+42] using features as known at the origin (no data leakage) without recursivity nor injection of predicted values.
- One row corresponds to: `(store, origine, cible)`.
- Features at origin (`asof_xxx`) correspond to last known state at origin and replicated on all 42 targets (i.e. the forecasting horizon).
- Feature `h` encodes how far the target is from the origin and in a way the freshness of the `asof_xxx` features from the target perspective
- KPI used: WMAPE, bias and FVA vs naive forecast by origin

### Results

Results of the model (without the `sales_ly` features):

| h_week | WMAPE | biais | FVA vs naïf |
|---|---|---|---|
| 1 | 12,9 % | +3,4 % | **−3,5 %** |
| 2 | 12,1 % | +4,7 % | +16,2 % |
| 3 | 10,9 % | +6,0 % | +33,5 % |
| 4 | 10,0 %  | +0,6 % | +31,0 % |
| 5 | 12,6 % | +1,1 % | +7,2 % |
| 6 | 10,1 % | +0,3 % | +9,1 % |


- **The model beats the seasonal naive one on the whole horizon except on week 1** for which the sales on the same day the previous year remains a better predictor (FVA -3.5%)
- **WMAPE does not get worse as the horizon increases.**
- **Bias is positive but remains low**. Bias at 6% on week 3 could be investigated.




### Key Learnings

`sales_ly` dominated the first model's feature importance (55% of the total while the second most important feature is at 10%) but removing it does not degrade the WMAPE (it even very slightly improves it). When removing it, the feature importance of `asof_mean_91d` grows by a factor of 2.6 while the feature importance of `asof_mean_28d` gets twice bigger. Furthermore, `asof_mean_91d` because the #1 feature.

Another interesting feature to look at is `asof_mean_7d` which went to a feature importance of 8500 to one of 118000 (x14) when `sales_ly` was removed. This short term signal was not useless, it was masked by the predominance of `sales_ly`.

The rigorous backtesting produced an honest WMAPE (~11.5%) and allowed us to measure the redundancy of the dominating feature, the falsely low importance of the short term signal, and the horizon on which the baseline is better than our model.


### Limits

There are slight discontinuity in the sales but 99.9% corresponds to normal closing days (weekends, sundays, public holidays) and have already been dealt with by filtering the dataset on open days only. There are 174 stores (out of 1115) which shows long closing period so in those case, the moving averages mix dates from before and after the interruption. This limitation is not addressed as we believe its effect is marginal.


## Test Run for MLFlow

### Wrapper

In [47]:
import mlflow
import matplotlib.pyplot as plt

mlflow.set_tracking_uri("sqlite:///mlflow.db")
mlflow.set_experiment("rossman_04_production")

def run_and_log(features, run_name):
    with mlflow.start_run(run_name=run_name):
        # ---- params ----
        mlflow.log_params({
            "n_features": len(features),
            "horizon": HORIZON,
            "n_test_origins": len(test_origins),
            "num_boost_round": 600,
            "has_sales_ly": "sales_ly" in features,
            **PARAMS,
        })
        mlflow.log_param("features", ", ".join(features))

        # ---- backtest (using the function already created) ----
        res, imp = run_backtest(features)

        # ---- metrics aggregated by horizon ----
        summary = res.groupby("h_week").apply(
            lambda g: pd.Series({
                "wmape": np.average(g["wmape"], weights=g["n"]),
                "bias":  np.average(g["bias"],  weights=g["n"]),
                "fva":   1 - np.average(g["wmape"], weights=g["n"])
                             / np.average(g["wmape_ly"], weights=g["n_ly"]),
            })
        )

        # one metrics by KPI and by step
        for hw, row in summary.iterrows():
            for kpi in ["wmape", "bias", "fva"]:
                mlflow.log_metric(kpi, float(row[kpi]), step=int(hw))

        # global aggregated metrics
        w = res.groupby("h_week")["n"].sum()
        mlflow.log_metric("wmape_global", float(np.average(summary["wmape"], weights=w)))
        mlflow.log_metric("bias_global",  float(np.average(summary["bias"],  weights=w)))

        # ---- feature importance as artefact and graphs ----
        imp_sorted = imp.sort_values(ascending=False)
        imp_sorted.to_csv("feature_importance.csv", header=["gain"])
        mlflow.log_artifact("feature_importance.csv")

        fig, ax = plt.subplots(figsize=(7, 5))
        imp_sorted.head(12).iloc[::-1].plot.barh(ax=ax)
        ax.set_xlabel("gain"); ax.set_title(f"Feature importance — {run_name}")
        fig.tight_layout()
        mlflow.log_figure(fig, "feature_importance.png")
        plt.close(fig)

        # ---- artefacts ----
        res.to_csv("backtest_detail.csv", index=False)
        summary.round(4).to_csv("summary_by_horizon.csv")
        mlflow.log_artifact("backtest_detail.csv")
        mlflow.log_artifact("summary_by_horizon.csv")

        return summary

### Runs

In [48]:
FEATURES_no_ly = [f for f in FEATURES if f != "sales_ly"]

summary_full  = run_and_log(FEATURES, "lgbm_full")
summary_no_ly = run_and_log(FEATURES_no_ly, "lgbm_no_sales_ly")

print("full :", summary_full["wmape"].round(4).to_dict())
print("no_ly:", summary_no_ly["wmape"].round(4).to_dict())

full : {1: 0.1337, 2: 0.1178, 3: 0.1187, 4: 0.0972, 5: 0.1202, 6: 0.1044}
no_ly: {1: 0.1294, 2: 0.1205, 3: 0.109, 4: 0.1002, 5: 0.1261, 6: 0.1013}
